pygmt/helpers/tempfile.py

```diff
@@ -126,18 +126,27 @@ def tempfile_from_geojson(geojson):
         E.g. '1a2b3c4d5e6.gmt'.
     """
     with GMTTempFile(suffix=".gmt") as tmpfile:
+        # pylint: disable=import-outside-toplevel
+        import geopandas as gpd
+
         os.remove(tmpfile.name)  # ensure file is deleted first
         ogrgmt_kwargs = {"filename": tmpfile.name, "driver": "OGR_GMT", "mode": "w"}
         try:
+            # Map int/int64 to int32 since OGR_GMT only supports 32-bit integer
+            # https://github.com/geopandas/geopandas/issues/967#issuecomment-842877704
+            # https://github.com/GenericMappingTools/pygmt/issues/2497
+            schema = gpd.io.file.infer_schema(geojson)
+            for col, dtype in schema["properties"].items():
+                if dtype in ("int", "int64"):
+                    schema["properties"][col] = "int32"
+            ogrgmt_kwargs["schema"] = schema
```

Comment:
Only write schema if schema["properties"] is not None. Should fix that error about `ValueError: Record does not match collection schema: KeysView(<fiona.model.Properties object at 0x7f3265f03a10>) != []`
```suggestion
            if schema["properties"]:
                ogrgmt_kwargs["schema"] = schema
```

Reply:
I think it's not the root cause of the error.

Here is a modified version of the `test_geopandas_info_geodataframe` test to understand why it fails:
```python
import geopandas as gpd
import shapely

linestring = shapely.geometry.LineString([(20, 15), (30, 15)])
polygon = shapely.geometry.Polygon([(20, 10), (23, 10), (23, 14), (20, 14)])
multipolygon = shapely.geometry.shape(
    {
        "type": "MultiPolygon",
        "coordinates": [
            [
                [[0, 0], [20, 0], [10, 20], [0, 0]],  # Counter-clockwise
                [[3, 2], [10, 16], [17, 2], [3, 2]],  # Clockwise
            ],
            [[[6, 4], [14, 4], [10, 12], [6, 4]]],  # Counter-clockwise
            [[[25, 5], [30, 10], [35, 5], [25, 5]]],
        ],
    }
)
# Multipolygon first so the OGR_GMT file has @GMULTIPOLYGON in the header
gdf = gpd.GeoDataFrame(
    index=["multipolygon", "polygon", "linestring"],
    geometry=[multipolygon, polygon, linestring],
)

schema = gpd.io.file.infer_schema(gdf)
print(gdf)
print(schema)

gdf.to_file("test.gmt", driver="OGR_GMT", mode="w")
```
The output of the two `print` statements is:
```
                                                       geometry
multipolygon  MULTIPOLYGON (((0.00000 0.00000, 20.00000 0.00...
polygon       POLYGON ((20.00000 10.00000, 23.00000 10.00000...
linestring    LINESTRING (20.00000 15.00000, 30.00000 15.00000)
{'geometry': ['MultiPolygon', 'Polygon', 'LineString'], 'properties': OrderedDict()}
```
To understand what geopandas/fiona does, I also added the same two print statements to source code `geopandas/io/file.py` at line https://github.com/geopandas/geopandas/blob/b4b10313ab57bf2c55592a28fb99687c9a538fc2/geopandas/io/file.py#L591.

The two print statements in the geopandas source code give:
```
          index                                           geometry
0  multipolygon  MULTIPOLYGON (((0.00000 0.00000, 20.00000 0.00...
1       polygon  POLYGON ((20.00000 10.00000, 23.00000 10.00000...
2    linestring  LINESTRING (20.00000 15.00000, 30.00000 15.00000)
{'geometry': ['MultiPolygon', 'Polygon', 'LineString'], 'properties': OrderedDict([('index', 'str')])}
```
As you can see, both the `gdf` object and the schema are different. This is because geopandas by defaults modifies the `index` column (see lines https://github.com/geopandas/geopandas/blob/b4b10313ab57bf2c55592a28fb99687c9a538fc2/geopandas/io/file.py#L551-L556).

In our case, our inferred schema has a empty property dict `OrderedDict()`, but geopandas/fiona expects a property dict like `OrderedDict([('index', 'str')])`.

Reply:
https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.to_file.html#geopandas.GeoDataFrame.to_file

> index: bool, default None
>
> If True, write index into one or more columns (for MultiIndex). Default None writes the index into one or more columns only if the index is named, is a MultiIndex, or has a non-integer data type. If False, no index is written.

I think writing `index` into one column or not has no effects to the output of `OGR_GMT` driver, so maybe we should just set `index=False`.

Reply:
> I think writing `index` into one column or not has no effects to the output of `OGR_GMT` driver, so maybe we should just set `index=False`.

Setting `index=True` (the default) will produce this in the output *.gmt file:

```
# @VGMT1.0
# @R0/35/0/20                                                             
# @GMULTIPOLYGON
# @Nindex
# @Tstring
# FEATURE_DATA
>
# @Dmultipolygon
# @P
0 0
20 0
10 20
0 0
>
# @H
3 2
10 16
17 2
3 2
>
# @P
6 4
14 4
10 12
6 4
>
# @P
25 5
30 10
35 5
25 5
>
# @Dpolygon
# @P
20 10
23 10
23 14
20 14
20 10
>
# @Dlinestring
20 15
30 15
```

Notice the lines with the comments `# @Nindex` and `# @Dmultipolygon` which are from the index. So the 'index' is output as just a regular column in the OGR_GMT file, and could potentially be used by someone doing `aspatial="Z=index"`. We probably do need to keep it just in case.

Is there another way to handle this? E.g. cast the 'index' column explicitly to a string type?

Reply:
> Is there another way to handle this? E.g. cast the 'index' column explicitly to a string type?

This won't always work, because the index property in a GeoDataFrame object can be integers. For example, in the above example, the object can be defined like
```
gdf = gpd.GeoDataFrame(
    #index=["multipolygon", "polygon", "linestring"],
    index=[0, 1, 2],
    geometry=[multipolygon, polygon, linestring],
)
```

Reply:
Maybe we should copy these lines from geopandas (https://github.com/geopandas/geopandas/blob/b4b10313ab57bf2c55592a28fb99687c9a538fc2/geopandas/io/file.py#L551-L556):
```
    if index is None:
        # Determine if index attribute(s) should be saved to file
        # (only if they are named or are non-integer)
        index = list(df.index.names) != [None] or not is_integer_dtype(df.index.dtype)
    if index:
        df = df.reset_index(drop=False)
```

Reply:
> ```
>     if index:
>         df = df.reset_index(drop=False)
> ```

Probably just need these two lines, call `reset_index` if an index exists.

Reply:
https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.to_file.html#geopandas.GeoDataFrame.to_file

`index` has a default value of `None`. So `df.reset_index` is only called if the GeoDataFrame object's index is named or is non-integer.

Reply:
Ah sorry, I thought `index` was referring to the `df.index` attribute. We could do something like this to force an index name:

```
if gdf.index.name is None:
    gdf.index.name = "index"
    # print(gdf.index.names) 
    # FrozenList(['index'])

```

Reply:
It looks good.

Reply:
Patch applied in 2097add02f44949f6668aa63539666c341d2aa66.